### PR TITLE
Add open_clip support

### PR DESCRIPTION
**Pull request by cat-state**
This PR adds support for open_clip to `clip_inference`. Models can be used by specifiying them as `clip_inference --clip_model "open_clip:<model_name>"`.

### CODE REVIEW

**Comment by rom1504** [comment]
You can get lint feedback by running `make black lint`

**Comment by rom1504** [comment]
Can you add info in the readme on how to use it ?

File: README.md [hunk 1]
```diff
@@ -88,7 +88,7 @@ clip_inference turn a set of text+image into clip embeddings
 * **write_batch_size** Write batch size (default *10**6*)
 * **wds_image_key** Key to use for images in webdataset. (default *jpg*)
 * **wds_caption_key** Key to use for captions in webdataset. (default *txt*)
-* **clip_model** CLIP model to load (default *ViT-B/32*)
+* **clip_model** CLIP model to load (default *ViT-B/32*). Specifiy it as `"open_clip:Vit-B/32-quickgelu"` to use the [open_clip](https://github.com/mlfoundations/open_clip) models.
```
Comment (rom1504) [inline README.md]: Can you also say here that to use open clip, it's needed to `pip install open_clip_torch` ?

We can also consider adding it directly in the dependencies if it doesn't transitively depend on too many things

Comment (cat-state) [reply README.md]: I'm not sure what requirements it has but the environment.yml in its repo is quite big.

**Comment by rom1504** [comment]
One more thing that can be useful: add open_clip_torch in test dependencies
Then add another test case in https://github.com/rom1504/clip-retrieval/blob/main/tests/test_clip_inference/test_mapper.py#L14

Test cases can be added by adding a decorator like this before the test function @pytest.mark.parametrize("model", ["ViT-B/32", "open_clip:ViT-B/32"])

File: README.md (hunk 1 above)
Comment (rom1504) [reply README.md]: The actual requirements of the pypi package are listed there https://github.com/mlfoundations/open_clip/blob/main/setup.py#L48

Comment (rom1504) [reply README.md]: It seems to me that means 0 new transitive dependencies
Worth checking by installing it in an environment where clip retrieval is already installed

File: tests/test_clip_inference/test_mapper.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 import pickle
 import os
+
 from clip_retrieval.clip_inference.mapper import ClipMapper
-import os
 
 
-def test_mapper():
+@pytest.mark.parametrize("model", ["ViT-B/32", "open_clip:ViT-B-32-quickgelu"])
+def test_mapper(model):
```
Comment (rom1504) [inline tests/test_clip_inference/test_mapper.py:9]: You need to pass that variable down there

Comment (cat-state) [reply tests/test_clip_inference/test_mapper.py:9]: oops

File: requirements-test.txt [hunk 3]
```diff
@@ -1,9 +1,10 @@
 img2dataset
 black==22.3.0
 mypy==0.942
+open-clip-torch==0.2.1
```
Comment (rom1504) [inline requirements-test.txt:4]: If you confirm that this adds no additional transitive dependencies then let's just make it a normal dependency (and remove the instruction to install it manually in the readme)

File: README.md (hunk 1 above)
Comment (cat-state) [reply README.md]: i think they have conflicting dependencies for webdataset as specified but it seems to work anyway

**Comment by rom1504** [comment]
https://github.com/mlfoundations/open_clip/blob/main/setup.py#L32 can you do a PR to add 3.6 to the list there ? That's why tests are failing

**Comment by rwightman** [comment]
@rom1504 @cat-state FYI, unlike OpenAI CLIP, OpenCLIP does not default to loading the models as scripted + fp16, the scripting has little impact in my tests so far, but I'd highly recommend running the OpenCLIP model under an amp autocast() context for increased throughput.

One should verify the retreival performance isn't impacted, but I've seen no impact in zero-shot validation scores comparing float32 and AMP

**Comment by rom1504** [comment]
blocker here is webdataset 0.2 used in openclip
can't use that here because webdataset 0.2 depends on torch and I don't want it in img2dataset https://github.com/rom1504/img2dataset/pull/163

**Comment by rom1504** [comment]
thanks @cat-state !

**Comment by rom1504** [comment]
tracking autocast in https://github.com/rom1504/clip-retrieval/issues/138

**Comment by rom1504** [comment]
let's also add openclip benchmark as part of https://github.com/rom1504/clip-retrieval/issues/116

**Pull request merged**